Get node unhealthy conditions from the ConfigMap

cynepco3hahue opened the pull request:
ConfigMap includes node unhealthy conditions with condition status and timeout. The code will check if the node has the specific unhealthy condition more than condition timeout and if yes will delete the machine object.

ConfigMap has next structure:
```
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    noderecovery.kubevirt.io: ""
  name: node-unhealth-conditions
  namespace: openshift-machine-api
data:
  conditions: |
    items:
    - name: Ready 
      timeout: 60s  # support all duration units "ns", "us" (or "µs"), "ms", "s", "m", "h"
      status: Unknown
```

enxebre commented on pkg/controller/machinehealthcheck/machinehealthcheck_controller.go:
```diff
@@ -170,23 +182,56 @@ func remediate(r *ReconcileMachineHealthCheck, machine *mapiv1.Machine) (reconci
+	err = r.client.Get(context.TODO(), cmKey, cmUnhealthConditions)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			glog.Warningf("ConfigMap %s not found", healthcheckingv1alpha1.ConfigMapNodeUnhealthConditions)
```
shall we state in the warning message that remediation won't happen if we end up here?

enxebre commented on pkg/controller/machinehealthcheck/machinehealthcheck_controller.go:
```diff
@@ -170,23 +182,56 @@ func remediate(r *ReconcileMachineHealthCheck, machine *mapiv1.Machine) (reconci
+
+			now := time.Now()
+			durationUnhealthy := now.Sub(nodeCondition.LastTransitionTime.Time)
 			glog.Warningf("machine %s have had node %s unhealthy for %s. Requeuing...", machine.Name, node.Name, durationUnhealthy.String())
```
nit: we could print the specific failing conditions here

enxebre commented on pkg/controller/machinehealthcheck/machinehealthcheck_controller.go:
```diff
@@ -170,23 +182,56 @@ func remediate(r *ReconcileMachineHealthCheck, machine *mapiv1.Machine) (reconci
+	var result *reconcile.Result
+	for _, c := range nodeUnhealthConditions {
+		nodeCondition := conditions.GetNodeCondition(node, corev1.NodeConditionType(c.Name))
+		if nodeCondition != nil && nodeCondition.Status == corev1.ConditionStatus(c.Status) {
```
nit: although it wouldn't give us reusability, I'd personally like to see this `if` sentence in an english readable format for clarity

enxebre commented:
there's a conflicting file `pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go`
@cynepco3hahue  Overall this looks great to me!!! thanks a lot!
We need to add an e2e test to validate that:
1 - The configMap is deployed 
2 - Given a machineHealthCheck resource is apply the config is consumed by the remediation logic even if remediation does not need to take action

PTAL @frobware @ingvagabund @bison

cynepco3hahue commented:
> there's a conflicting file `pkg/controller/machinehealthcheck/machinehealthcheck_controller_test.go`
> @cynepco3hahue Overall this looks great to me!!! thanks a lot!
> We need to add an e2e test to validate that:
> 1 - The configMap is deployed
> 2 - Given a machineHealthCheck resource is apply the config is consumed by the remediation logic even if remediation does not need to take action

@enxebre I rebased and add the first e2e test, but I do not sure how can I add the second one, does openshift have some utilities to run commands on the node(like `systemctl stop kubelet`) because I want to add testing for the whole machine health check flow, similar to one that I had under `node-recovery` [repository](https://github.com/kubevirt/node-recovery/blob/master/tests/node_remediation_test.go).

enxebre commented:
/retest

cynepco3hahue commented:
/retest

enxebre commented:
/ok-to-test

enxebre commented:
/retest

cynepco3hahue commented:
@enxebre Looks like some problems with the CI
```
error: unable to parse image registry.svc.ci.openshift.org/ci-op-yw7v3b8s/stable@sha256:afa3de170f458beb5af5b45eadc129179d0bbec77b62cdde45d02c7ea84560c3: cannot retrieve image configuration for manifest sha256:afa3de170f458beb5af5b45eadc129179d0bbec77b62cdde45d02c7ea84560c3: received unexpected HTTP status: 504 Gateway Time-out
```

cynepco3hahue commented:
/retest

cynepco3hahue commented:
@enxebre Do you why the CI is failing? I see a lot of
`level=warning msg="RetryWatcher - getting event failed! Re-creating the watcher. Last RV: 185"` looks like the installer problem.

cynepco3hahue commented:
/ok-to-test

cynepco3hahue commented:
@enxebre @ingvagabund I created machine health check e2e smoke tests.
Currently, I created it under the vendor directory, but when this PR will be merged I will move them to https://github.com/openshift/cluster-api-actuator-pkg and will update `Gopkg.toml` to consume them.
Do I understand the flow correct?

cynepco3hahue commented:
/retest

ingvagabund commented:
> Currently, I created it under the vendor directory, but when this PR will be merged I will move them to https://github.com/openshift/cluster-api-actuator-pkg and will update `Gopkg.toml` to consume them.
Do I understand the flow correct?

We are still discussing the right approach thought this is the best we right now have :)

ingvagabund commented:
@cynepco3hahue sorry for the mess, `e2e/openshift` is now deprecated and being removed. The right way from now on is to update tests under `vendor/github.com/openshift/cluster-api-actuator-pkg/pkg/e2e` of one of the dirs. In this case the most suitable package is `infra`.

cynepco3hahue commented:
@ingvagabund Oh it just passed for me :) Will align it with the update.

My 2 cents about ginkgo:

- you can use
```
. "github.com/onsi/ginkgo"
. "github.com/onsi/gomega"
```
it will make test code more clear

- you can use `Eventually` that more TDD way than `wait.PollImmediate`

enxebre commented:
@cynepco3hahue this is looking great! sorry you got us in the middle of transitioning the e2e testing could you move the e2e validation to https://github.com/openshift/machine-api-operator/tree/master/vendor/github.com/openshift/cluster-api-actuator-pkg/pkg/e2e/infra and add the counter part PR here https://github.com/openshift/cluster-api-actuator-pkg/?

cynepco3hahue commented:
@enxebre I updated the PR to use changes from 
`https://github.com/openshift/cluster-api-actuator-pkg/`
- added new `machinehealthcheck` suite, that we can extend in the future
- added some help methods

I want to create the PR under the https://github.com/openshift/cluster-api-actuator-pkg/, after that this one will be merged because this PR has some constants and methods that I call from tests.

cynepco3hahue commented:
@enxebre @ingvagabund @beekhof Guys can you please take a look? I hope CI will pass, and IMHO it is ready for the final review(before I need to refactor it again :sweat_smile: )

cynepco3hahue commented:
/retest

enxebre commented:
/retest

cynepco3hahue commented:
ci test please

ingvagabund commented:
@cynepco3hahue I see you are changing the controller loop logic. Who would be creating/updating content of the configmap in AWS now? Is there any external controller who does that?

cynepco3hahue commented:
@ingvagabund Do not sure if I understand your question, can you please clarify?

ingvagabund commented:
- if the CM is not found, error [1]
- you are deciding if a node is unhealthy for long based on the CF content [2]

Right now, we pull all the nodes and check their ready condition. With this PR, there is noone updating the CM that carries node status of all the nodes in the cluster. So when we deploy the machine health check controller as it is in this PR, it will not work since noone is updating the CF.

You are providing the CF here [3]. Who/what is updating `data.conditions` field?

[1] https://github.com/openshift/machine-api-operator/pull/204/commits/d07898c39af5fd3edfaaa139c7ddcfca24e22e41#diff-614d58186947ca2e4e215d42c496d72eR191
[2] https://github.com/openshift/machine-api-operator/pull/204/commits/d07898c39af5fd3edfaaa139c7ddcfca24e22e41#diff-614d58186947ca2e4e215d42c496d72eR203
[3] https://github.com/openshift/machine-api-operator/pull/204/commits/532b1c387f8544ad86f015ae7da4716ba2998157#diff-4464215dee322b5ea7ff7872e66318c8

cynepco3hahue commented:
@ivanhorvath I believe the user should update conditions because it can be a lot of different factors, also in the case when NPD(node problem detector) installed the user can provide more granular conditions.
But I can update the https://github.com/openshift/machine-api-operator/commit/532b1c387f8544ad86f015ae7da4716ba2998157#diff-4464215dee322b5ea7ff7872e66318c8 to include `Ready` conditions with type `Unknown` with some reasonable timeout, just give me know what do you prefer?

beekhof commented:
I would suggest that the old NodeReady/5min condition should be used if items[] is empty, or the CM is completely not available

cynepco3hahue commented:
@beekhof I prefer do not have any hard-coded requirements for remediation under the code because each one can have a different timeout for remediation. But I will add a default value to the config map that the operator deploy.

But I still in trouble with e2e tests, I checked to stop `hyperkube` process on the libvirt environment and all works as expected, so I do not really understand why it does not work for AWS environment.

cynepco3hahue commented:
/retest

cynepco3hahue commented:
@enxebre @ingvagabund Guys can you please review, `e2e-aws-operator` tests passed, `integration` tests cannot pass, because it uses dummy nodes and not real ones.

ingvagabund commented:
> because it uses dummy nodes and not real ones.

@cynepco3hahue kubemark provider config has `turnUnhealthyAfter` and `healthyDuration` fields which you can use to tell a node to start reporting NotReady. E.g.
```
apiVersion: kubemarkproviderconfig.k8s.io/v1alpha1
kind: KubemarkMachineProviderConfig
healthyDuration: 40s
turnUnhealthyAfter: true
```

You just need to create your own machine manifest for kubemark.

cynepco3hahue commented:
@ingvagabund I am sure I can, but it does not the single thing

- what will happen when I will create pod that should stop kubelet on the dummy node
- does the dummy node has `Ready` condition with the status False or Unknown(in most problematic cases I saw that it `Unknown`)

I can introduce some condition logic under tests, but I really do not want to do it, because it will make things more complicated.

And another question, what value does it provide to run the same tests that we ran vs real environment on the environment with dummy nodes?
It can be useful to use dummy nodes because it provides the way to run tests much faster(do not need to wait for the `Ready` condition with the status `Unknown`, do not need to wait for provisioning of the new node, after the machine deletion), but I think it should be a different test suite.

Any thought?

ingvagabund commented:
@ingvagabund I am sure I can, but it does not the single thing

> what will happen when I will create pod that should stop kubelet on the dummy node
does the dummy node has Ready condition with the status False or Unknown(in most problematic cases I saw that it Unknown)

Any pod scheduled to a dummy node gets into `Running` state

> I can introduce some condition logic under tests, but I really do not want to do it, because it will make things more complicated.

Sometimes there is no other way that to create specific implementation per cloud provider.

> And another question, what value does it provide to run the same tests that we ran vs real environment on the environment with dummy nodes?
It can be useful to use dummy nodes because it provides the way to run tests much faster(do not need to wait for the Ready condition with the status Unknown, do not need to wait for provisioning of the new node, after the machine deletion), but I think it should be a different test suite.

To design tests that are actuator independent. The machine health-check controller is cloud provider free. The controller is allowed to access node and other kubernetes objects. Thought, it's not allowed to ssh to nodes, to make decision based on cloud provider infrastructure, etc.

cynepco3hahue commented:
@ingvagabund Guys you did whole development process such complicate...
So to make it work with `KubemarkMachineProviderConfig` I will need to update(or create) the machine object via API, so I want to bring `KubemarkMachineProviderConfig` API, does it ok?

cynepco3hahue commented:
/retest

cynepco3hahue commented:
/retest

cynepco3hahue commented:
@enxebre @ingvagabund Can you guys please take a look?
And currently, I skip machine health check tests for the kubemark provider.
Maybe in the future, we can tweak somehow kubemark provider to have the possibility to make it `NotReady` by some update of configmap and without update machine object.

ingvagabund commented on vendor/github.com/openshift/cluster-api-actuator-pkg/pkg/e2e/machinehealthcheck/machinehealthcheck.go:
```diff
@@ -0,0 +1,217 @@
+			Expect(err).ToNot(HaveOccurred())
+		}
+
+		listOptions.SetLabelSelector(fmt.Sprintf("%s=", e2e.KubeletKillerPodName))
```
Can you moved 206-215 under a function?

cynepco3hahue replied:
Done

ingvagabund commented on vendor/github.com/openshift/cluster-api-actuator-pkg/pkg/e2e/machinehealthcheck/machinehealthcheck.go:
```diff
@@ -0,0 +1,217 @@
+		listOptions.SetLabelSelector(fmt.Sprintf("%s=", e2e.NodeWorkerLabel))
+		workers := &corev1.NodeList{}
+		glog.V(2).Infof("Wait until the environment will have %d ready workers", numberOfReadyWorkers)
+		Eventually(func() bool {
```
Can you move 175-191 under a function?

cynepco3hahue replied:
done

ingvagabund commented on vendor/github.com/openshift/cluster-api-actuator-pkg/pkg/e2e/machinehealthcheck/machinehealthcheck.go:
```diff
@@ -0,0 +1,217 @@
+
+	waitNodeUnhealthyCondition := func() {
+		key := types.NamespacedName{
+			Name:      workerNode.Name,
```
Can you make this (`workerNode.Name`) an argument of waitNodeUnhealthyCondition function? So there are no "magic" values read outside of well-defined function signature. It's even fine to move `waitNodeUnhealthyCondition` out of `Describe`.

ingvagabund commented on pkg/controller/machinehealthcheck/machinehealthcheck_controller.go:
```diff
@@ -170,27 +182,82 @@ func remediate(r *ReconcileMachineHealthCheck, machine *mapiv1.Machine) (reconci
+	}
+
+	var result *reconcile.Result
+	for _, c := range nodeUnhealthConditions {
```
Default deployments will always have either empty CM or no CM. In the first case, the machine controller will not do anything until the CM is updated with valid data by third party controller. In the second case, the machine controller will fall into failed-all-the-time loop. Is this the expected behavior? @enxebre WDYT?

ingvagabund replied:
I see, I did not see it as it was hidden in `getUnhealthConditionsConfigMap` call. Ok.

ingvagabund commented:
There are other lines of code (I did not mention)  worth moving under a function/method to make the tests more readable.

cynepco3hahue commented on vendor/github.com/openshift/cluster-api-actuator-pkg/pkg/e2e/machinehealthcheck/machinehealthcheck.go:
```diff
@@ -0,0 +1,217 @@
+		})
+
+		It("should delete unhealthy machine", func() {
+			By(fmt.Sprintf("Stopping kubelet service on the node %s", workerNode.Name))
```
done

cynepco3hahue commented on vendor/github.com/openshift/cluster-api-actuator-pkg/pkg/e2e/machinehealthcheck/machinehealthcheck.go:
```diff
@@ -0,0 +1,217 @@
+				}
+			}
+			return false
+		}, 360*time.Second, 5*time.Second).Should(BeTrue())
```
done

cynepco3hahue commented on pkg/controller/machinehealthcheck/machinehealthcheck_controller.go:
```diff
@@ -170,27 +182,82 @@ func remediate(r *ReconcileMachineHealthCheck, machine *mapiv1.Machine) (reconci
+			nodeCondition.Type,
+			durationUnhealthy.String(),
+		)
+		result = &reconcile.Result{Requeue: true}
```
@ingvagabund Hm when you have a different timeout for each condition you do not really want to wait always 5 minutes.
I can change it to `result = &reconcile.Result{Requeue: true, RequeueAfter: conditionTimeout}, will it work for you?

enxebre commented on pkg/controller/machinehealthcheck/machinehealthcheck_controller.go:
```diff
@@ -170,27 +186,92 @@ func remediate(r *ReconcileMachineHealthCheck, machine *mapiv1.Machine) (reconci
+			nodeCondition.Type,
+			durationUnhealthy.String(),
+		)
+		result = &reconcile.Result{Requeue: true, RequeueAfter: conditionTimeout}
```
is `var result` only used here?
Why not `return reconcile.Result{Requeue: true, RequeueAfter: conditionTimeout}, nil`

cynepco3hahue replied:
@enxebre The problem that node can have the number of conditions, and only one of them is last long enough, so I do not want to return from the loop.

ingvagabund replied:
If that's true you should than take minimum of all `conditionTimeout` values.

ingvagabund commented on pkg/controller/machinehealthcheck/machinehealthcheck_controller.go:
```diff
@@ -170,27 +186,92 @@ func remediate(r *ReconcileMachineHealthCheck, machine *mapiv1.Machine) (reconci
+		}
+
+		// creates dummy config map with default values if it does not exist
+		cmUnhealthConditions, err = conditions.CreateDummyUnhealthyConditionsConfigMap()
```
@enxebre this needs to be documented though so a user is not surprised in case he/she deletes the configmap (or all producers of it)

cynepco3hahue replied:
@ingvagabund it is another question, do I need to add documentation somewhere?

enxebre replied:
@cynepco3hahue can you include it here https://github.com/openshift/machine-api-operator#machine-healthcheck-controller

enxebre commented on pkg/controller/machinehealthcheck/machinehealthcheck_controller.go:
```diff
@@ -170,27 +186,92 @@ func remediate(r *ReconcileMachineHealthCheck, machine *mapiv1.Machine) (reconci
+		if err != nil {
+			return nil, err
+		}
+		glog.Warningf(
```
No need to warn the user on our default case. Info would be ok

cynepco3hahue replied:
done

enxebre commented on pkg/controller/machinehealthcheck/machinehealthcheck_controller.go:
```diff
@@ -170,27 +186,92 @@ func remediate(r *ReconcileMachineHealthCheck, machine *mapiv1.Machine) (reconci
+	return cmUnhealthConditions, nil
+}
+
+func isConditionsStatusesEqual(cond *corev1.NodeCondition, unhealthCond *conditions.UnhealthyCondition) bool {
```
we have `conditions.UnhealthyCondition` but name the variable `unhealthCond` can we use `unhealthy` consistently?

cynepco3hahue replied:
done

ingvagabund commented on vendor/github.com/openshift/cluster-api-actuator-pkg/pkg/e2e/machinehealthcheck/machinehealthcheck.go:
```diff
@@ -0,0 +1,225 @@
+	}, e2e.WaitLong, 10*time.Second).Should(BeTrue())
+}
+
+func waitForReadyWorkers(numberOfReadyWorkers int) {
```
s/waitForReadyWorkers/waitForWorkersToGetReady`

ingvagabund commented on vendor/github.com/openshift/cluster-api-actuator-pkg/pkg/e2e/machinehealthcheck/machinehealthcheck.go:
```diff
@@ -0,0 +1,225 @@
+	})
+})
+
+func waitNodeUnhealthyCondition(workerNodeName string) {
```
`s/waitNodeUnhealthyCondition/waitForNodeUnhealthyCondition`

cynepco3hahue replied:
done

ingvagabund commented on vendor/github.com/openshift/cluster-api-actuator-pkg/pkg/e2e/machinehealthcheck/machinehealthcheck.go:
```diff
@@ -0,0 +1,225 @@
+	}, e2e.WaitLong, 10*time.Second).Should(BeTrue())
+}
+
+func waitForReadyWorkers(numberOfReadyWorkers int) {
```
`s/waitForReadyWorkers/waitForWorkersToGetReady`

cynepco3hahue replied:
done

ingvagabund commented on vendor/github.com/openshift/cluster-api-actuator-pkg/pkg/e2e/machinehealthcheck/machinehealthcheck.go:
```diff
@@ -0,0 +1,225 @@
+
+	stopKubeletAndValidateMachineDeletion := func(workerNode *corev1.Node, workerMachine *mapiv1beta1.Machine, timeout time.Duration) {
+		By(fmt.Sprintf("Stopping kubelet service on the node %s", workerNode.Name))
+		err := e2e.StopKubelet(workerNode.Labels["kubernetes.io/hostname"])
```
Node hostname does not always has to be equal to node name. Plus, you can use `workerNode.Name` directly.

cynepco3hahue replied:
under the pod creation I use
```
NodeSelector: map[string]string{
      "kubernetes.io/hostname": nodeName,
},
```
so it for sure will use the right node

ingvagabund replied:
You can set `.Spec.NodeName` directly to `nodeName`.

enxebre commented on pkg/util/conditions/conditions_test.go:
```diff
@@ -0,0 +1,196 @@
+						Name:    "Ready",
+						Timeout: "60s",
+						Status:  "Unknown",
+					},
```
why unknown? what about false?

cynepco3hahue replied:
I do not sure when the kubelet get `False` condition, in most problematic cases that I saw, kubelet just stop to talk with the master, so the node has `Unknown` status.

ingvagabund commented on vendor/github.com/openshift/cluster-api-actuator-pkg/pkg/e2e/machinehealthcheck/machinehealthcheck.go:
```diff
@@ -0,0 +1,217 @@
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	Context("with node-unhealth-conditions configmap", func() {
```
So you are using `Context` to have extra `BeforeEach` and `AfterEach`.

enxebre commented on pkg/apis/healthchecking/v1alpha1/machinehealthcheck_types.go:
```diff
@@ -4,6 +4,9 @@ import (
 )
 
+// ConfigMapNodeUnhealthConditions contains the name of the unhealth conditions config map
+const ConfigMapNodeUnhealthConditions = "node-unhealth-conditions"
```
can we use unhealthy instead consistently?

cynepco3hahue replied:
done

ingvagabund commented on vendor/github.com/openshift/cluster-api-actuator-pkg/pkg/e2e/machinehealthcheck/machinehealthcheck.go:
```diff
@@ -0,0 +1,225 @@
+	}, 15*time.Minute, 10*time.Second).Should(BeTrue())
+}
+
+func deleteMachineHealthChecks() {
```
Deleting all HC resources might cause trouble when we start running multiple tests in parallel. Can you delete only the single HC that was created? `CreateMachineHealthCheck` is creating only one HC. So `deleteMachineHealthChecks` as its counterpart should then delete only the same HC resource that was created.

cynepco3hahue replied:
done

enxebre commented on pkg/util/conditions/conditions_test.go:
```diff
@@ -0,0 +1,196 @@
+)
+
+const (
+	namespace   = "openshift-cluster-api"
```
nit: openshift-machine-api

cynepco3hahue replied:
done

enxebre commented on pkg/util/conditions/conditions.go:
```diff
@@ -0,0 +1,74 @@
+		Items: []UnhealthyCondition{
+			{
+				Name:    "Ready",
+				Status:  "Unknown",
```
why unknown? what about false?

cynepco3hahue replied:
I added False condition as well, but again I do not sure when it appears at all.

ingvagabund commented on pkg/controller/machinehealthcheck/machinehealthcheck_controller.go:
```diff
@@ -170,27 +186,92 @@ func remediate(r *ReconcileMachineHealthCheck, machine *mapiv1.Machine) (reconci
+		now := time.Now()
+		durationUnhealthy := now.Sub(nodeCondition.LastTransitionTime.Time)
+		glog.Warningf(
+			"machine %s have unhealthy node %s with condition %s for %s. Requeuing...",
```
Can you print the timeout as well? So a user knows how much time is left before a node gets removed?

cynepco3hahue replied:
done

cynepco3hahue commented on pkg/controller/machinehealthcheck/machinehealthcheck_controller.go:
```diff
@@ -170,27 +186,92 @@ func remediate(r *ReconcileMachineHealthCheck, machine *mapiv1.Machine) (reconci
+		return reconcile.Result{}, err
+	}
+
+	nodeUnhealthConditions, err := conditions.GetNodeUnhealthyConditions(node, cmUnhealthConditions)
```
done

ingvagabund commented on vendor/github.com/openshift/cluster-api-actuator-pkg/pkg/e2e/framework/machinehealthcheck.go:
```diff
@@ -0,0 +1,123 @@
+			},
+			HostPID: true,
+			NodeSelector: map[string]string{
+				"kubernetes.io/hostname": nodeName,
```
You can set `.Spec.NodeName` directly to `nodeName`.

cynepco3hahue commented:
@ingvagabund @enxebre I fixed all comments and added a section under the `README` I hope I did not miss any comment and thanks a lot of the review.

ingvagabund commented on README.md:
```diff
@@ -80,17 +80,33 @@ $ make nodelink-controller
+  
   1. Pick a node that is managed by one of the machineset's machines
-  1. SSH into the node, disable and stop the kubelet services:
+  2. SSH into the node, disable and stop the kubelet services:
```
nit: you can keep the numbering as it is. It will gets automatically increased.

cynepco3hahue replied:
done

ingvagabund commented on README.md:
````diff
@@ -80,17 +80,33 @@ $ make nodelink-controller
      ```
 
+  1. By default, the machine health check controller recognize only `NotReady` condition and will remove
+     unhealthy machine after 5 minutes, if you want to customize unhealthy conditions you can create `node-unhealthy-conditions` config map, for example:
````
nit: s/, if/. If/

cynepco3hahue replied:
done

ingvagabund commented:
/approve

cynepco3hahue commented:
@enxebre Can you please review again?

ingvagabund commented:
/lgtm
